library/core/src/intrinsics.rs

```diff
@@ -2666,12 +2666,17 @@ extern "rust-intrinsic" {
     ///
     /// `catch_fn` must not unwind.
     ///
-    /// The third argument is a function called if an unwind occurs (both Rust unwinds and foreign
-    /// unwinds). This function takes the data pointer and a pointer to the target-specific
-    /// exception object that was caught. For more information, see the compiler's source as well as
-    /// std's `catch_unwind` implementation.
+    /// The third argument is a function called if an unwind occurs (both Rust `panic` and foreign
+    /// unwinds). This function takes the data pointer and a pointer to the target- and
+    /// runtime-specific exception object that was caught.
     ///
-    /// The stable version of this intrinsic is `std::panic::catch_unwind`.
+    /// Note that in the case of a foreign unwinding operation, the exception object data may not be
+    /// safely usable from Rust, and should not be directly exposed via the standard library. To
+    /// prevent unsafe access, the library implementation may either abort the process or present an
+    /// opaque error type to the user.
```

Comment:
This sounds like it is the library implementation that chooses which of the two happen, which I don't think is the case?

It would be good to explain somewhere (not as stable docs, but as information for people that want to understand the system) what decides which of these two outcomes occurs.

Reply:
I thought that the library did indeed make that decision from my first look at the code, but on looking again I'm not so sure.

The files where the detection, and handling, of "foreign" exceptions occur seems to be in `rust/library/panic_unwind/src/<target>.rs`.

@workingjubilee or @chorman0773 , can either of you confirm that this understanding is correct, and that the standard library, rather than the compiler intrinsic, determines whether `catch_unwind` aborts in the face of a foreign exception?

Reply:
I can't speak to how rustc in particular handles things, I can only speak to abi-level handling and guarantees.

Reply:
@nbdd0121 Are you the right person to ask here?

Reply:
Yeah, the library decides

Reply:
Hi, sorry. Yes, that's correct, we call `__rust_foreign_exception`, which shows up in these places:
```
panic_unwind/src/lib.rs
84:    fn __rust_foreign_exception() -> !;

panic_unwind/src/seh.rs
366:        super::__rust_foreign_exception();
372:        super::__rust_foreign_exception();

panic_unwind/src/gcc.rs
89:        super::__rust_foreign_exception();
101:        super::__rust_foreign_exception();

panic_unwind/src/emcc.rs
78:        super::__rust_foreign_exception();
83:        super::__rust_foreign_exception();

std/src/panicking.rs
64:extern "C" fn __rust_foreign_exception() -> ! {
```

Reply:
Our unwinding code is unfortunately not very localized, and is splashed all over the library folder (workspace, now?). Some of the relevant code is inside library/std and some of it is in library/panic_unwind and some of it is in library/unwind and just... ugh.

I think half the reason this situation has not been cleaned up is because it's hard to just _find it all_.

Reply:
That sounds like we currently always abort (seeing as `__rust_foreign_exception` has return type `!`)? But the point of the docs here is that we may change that in the future?

Reply:
Yes, we currently always abort, which is why no implementation change accompanies these doc changes. The point of the doc change here is to _constrain_ what we may do in the future, because it is currently officially UB.